### ISSUE TITLE
FAbs and flushing denorms

### ISSUE THREAD

**Issue by jholewinski** [Microsoft/DirectXShaderCompiler#236]
ExecutionTest::UnaryFloatOpTest#FAbs is expecting denorms to be preserved:

```xml
        <Parameter Name="Validation.Input">
          ...
          <Value>-denorm</Value>
          ...
        </Parameter>
        <Parameter Name="Validation.Expected">
          ...
          <Value>denorm</Value>
          ...
        </Parameter>
```

The DXIL spec does not mention whether denorms must be preserved for FAbs, though most operations do specify flush to zero behavior.

**Comment by youngkim93** [comment]
We probably want to keep this behavior as this is what DXBC also did with its 'abs' operations. I will update the spec to address this.

**Comment by jholewinski** [comment]
Where does the HLSL/FXC spec require this? According to the ["Floating-Point Rules" document](https://msdn.microsoft.com/en-us/library/windows/desktop/cc308050(v=vs.85).aspx), it seems the implementation must flush to zero:

```
Denorms are flushed to sign-preserved zero on input and output of any floating-point mathematical operation. 
```

The [documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/hh446814(v=vs.85).aspx) for the `_abs` modifier only says that the sign is forced to positive, and does not mention denorms. That seems to imply that the first rule above is in effect (implementation must flush to zero) since absolute value is a "mathematical operation."

**Comment by tex3d** [comment]
In DXBC, the abs on a float is implemented as an input operand modifier on a mov instruction, which isn't considered a math op.  It is not used for int because mov is input type agnostic.  It should simply clear the high bit as far as I know.  For DXIL the FAbs is a separate op, but the behavior should match DXBC to maintain compatibility so people don't have to change HLSL code to bitcast to int and clear the high bit and bitcast back.

Also, the documentation for _abs is a holdover from DX9 with DX10 targets tacked on.  It looks like this in DX10+ DXBC:
mov o0.xyzw, |v0.xyzw|

This modifier may be used on an operand of a non-math op as well.  What is the interpretation in that case?

What do drivers do with the above instruction today?

**Comment by youngkim93** [comment]
This issue should be resolved with #249 to allow denorms to flush to sign-preserved zero. Please let us know if there is still an issue with this test case. Thanks!

**Comment by jholewinski** [comment]
Tests look good! Thanks!
